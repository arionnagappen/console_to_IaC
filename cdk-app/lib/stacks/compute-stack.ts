import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { MyLaunchTemp } from '../constructs/compute/launchTemp-construct';
import { MyAutoScalingGroup } from '../constructs/compute/asg-construct';
import { AppSecurityGroup } from '../constructs/compute/launchTemp-SG';
import * as ec2 from 'aws-cdk-lib/aws-ec2'

interface ComputeStackProps extends cdk.StackProps{
  vpc: ec2.IVpc
}

export class ComputeStack extends cdk.Stack {
  public readonly launchTempSg: ec2.ISecurityGroup;

  constructor(scope: Construct, id: string, props: ComputeStackProps){
    super(scope, id, props)

    const launchTempSg = new AppSecurityGroup(this, 'AppSG', {
      vpc: props.vpc
    });

    const launchTemp = new MyLaunchTemp(this, 'AsgLaunchTemp', {
      launchTempName: 'AppServerLaunchTemp',
      imageId: 'ami-0914bddde8faa93a0',
      instanceType: 't3.micro',
      securityGroupId: launchTempSg.securityGroup.securityGroupId
    });

    this.launchTempSg = launchTempSg.securityGroup;

    new MyAutoScalingGroup(this, 'AppServerASG', {
      vpc: props.vpc,
      launchTemplate: launchTemp.launchTemp,
      subnetType: ec2.SubnetType.PUBLIC,
      minSize: 1,
      maxSize: 3,
      desiredCapacity: 2,
    });
  };
};