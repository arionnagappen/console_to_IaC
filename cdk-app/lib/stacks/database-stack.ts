import { Construct } from "constructs";
import { DatabaseSecurityGroup } from "../constructs/database/rds-sg-construct";
import { MyRDSInstance } from "../constructs/database/rds-instance-construct"; 
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as rds from 'aws-cdk-lib/aws-rds';
import * as secretsmanager from 'aws-cdk-lib/aws-secretsmanager'
import * as cdk from 'aws-cdk-lib';

interface DBStackProps extends cdk.StackProps{
  vpc: ec2.IVpc
  appSecurityGroupId: ec2.ISecurityGroup["securityGroupId"];
}

export class DBStack extends cdk.Stack{
  constructor(scope: Construct, id: string, props: DBStackProps) {
    super(scope, id, props)

    // Security Group
    const dbSecurityGroup = new DatabaseSecurityGroup(this, 'DBSecurityGroup', {
      vpc: props.vpc,
      appSecurityGroupId: props.appSecurityGroupId
    });

    const mySubnetGroup = new rds.SubnetGroup(this, 'MyDbSubnetGroup', {
      description: 'Subnet group for RDS',
      vpc: props.vpc,
      vpcSubnets: {
        subnetType: ec2.SubnetType.PRIVATE_ISOLATED, 
      },
      removalPolicy: cdk.RemovalPolicy.DESTROY
    });

    // Retrieve Secret
    const dbSecret = secretsmanager.Secret.fromSecretNameV2(this, 'SecretName','rds-credentials');

    // Tell RDS use the Secret
    const myCredentials = rds.Credentials.fromSecret(dbSecret);

    new MyRDSInstance(this, 'MyDatabase', {
      vpc: props.vpc,
      dbSubnetGroup: mySubnetGroup,
      dbName: 'myDatabase',
      dbInstanceType: ec2.InstanceType.of(ec2.InstanceClass.T3, ec2.InstanceSize.MICRO),
      dbStorage: 10,
      dbEngine: rds.DatabaseInstanceEngine.mysql({version: rds.MysqlEngineVersion.VER_8_4_5}),
      dbCredentials: myCredentials,
      dbSecurityGroup: [dbSecurityGroup.dbSecurityGroup],
      dbMultiAZ: true,
    })

  }
}