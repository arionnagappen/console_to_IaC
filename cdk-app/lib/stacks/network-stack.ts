import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { MyVPC } from '../constructs/networking/my-vpc-construct';

export class NetworkStack extends cdk.Stack{
  constructor(scope: Construct, id: string, props?: cdk.StackProps){
    super(scope, id, props);

    const myVpc = new MyVPC(this, 'MyVpc', {
      cidr: '10.0.0.0/16',
      maxAzs: 2
    });

    new cdk.CfnOutput(this, 'VpcId', {
      value: myVpc.vpc.vpcId,
      description: 'VPC ID'
    });
  }
}