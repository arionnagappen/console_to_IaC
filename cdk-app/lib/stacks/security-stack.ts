import { IAMGroup } from "../constructs/security/iam/iam-group/iam_groups";
import { DeveloperPolicy } from "../constructs/security/iam/group-policies/dev-policies"
import { OperationsPolicy } from "../constructs/security/iam/group-policies/ops-policies";
import { FinancePolicy } from "../constructs/security/iam/group-policies/fin-policies";
import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";

interface IAMStackProps extends cdk.StackProps {}

export class IAMStack extends cdk.Stack{
  constructor(scope: Construct, id: string) {
    super(scope, id)

    // Policies
    const devPolicy = new DeveloperPolicy(this, 'DevPolicy');
    const opsPolicy = new OperationsPolicy(this, 'OperationsPolicy');
    const financePolicy = new FinancePolicy(this, 'FinancePolicy');

    // IAM Groups
    const devGroup = new IAMGroup(this, 'DeveloperGroup', {
      groupName: 'Developer-Group'
    });

    const opsGroup = new IAMGroup(this, 'OperationsGroup', {
      groupName: 'Operations-Group'
    });

    const financeGroup = new IAMGroup(this, 'FinanceGroup', {
      groupName: 'Finance-Group'
    });

    // Attach Policies
    devPolicy.attachToGroup(devGroup.iam_group);
    opsPolicy.attachToGroup(opsGroup.iam_group);
    financePolicy.attachToGroup(financeGroup.iam_group);
  }
}