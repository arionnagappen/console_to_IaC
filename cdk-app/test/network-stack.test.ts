import { App } from 'aws-cdk-lib'
import { Template } from 'aws-cdk-lib/assertions'
import { NetworkStack } from '../lib/stacks/network-stack'

// Assertion Test
test('VPC is created with correct CIDR', ()=> {
  const app = new App();
  const stack = new NetworkStack(app, 'TestStack');

  const template = Template.fromStack(stack);

  template.hasResourceProperties('AWS::EC2::VPC', {
    CidrBlock: '10.0.0.0/16'
  });
});

// Snapshot Test
test('VPC stack matches snapshot', ()=>{
  const app = new App();
  const stack = new NetworkStack(app, 'TestStack');

  const template = Template.fromStack(stack).toJSON();

  expect(template).toMatchSnapshot();
})
