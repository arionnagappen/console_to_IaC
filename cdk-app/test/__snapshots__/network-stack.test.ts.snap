// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VPC stack matches snapshot 1`] = `
{
  "Outputs": {
    "VpcId": {
      "Description": "VPC ID",
      "Value": {
        "Ref": "MyVpcMyVPC28A4F3FB",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "MyVpcMyVPC28A4F3FB": {
      "Properties": {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/MyVpc/MyVPC",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "MyVpcMyVPCIGW6C595E3B": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/MyVpc/MyVPC",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "MyVpcMyVPCVPCGW75395802": {
      "Properties": {
        "InternetGatewayId": {
          "Ref": "MyVpcMyVPCIGW6C595E3B",
        },
        "VpcId": {
          "Ref": "MyVpcMyVPC28A4F3FB",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
    "MyVpcMyVPCapplicationsubnetSubnet1DefaultRoute44E03D22": {
      "DependsOn": [
        "MyVpcMyVPCVPCGW75395802",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "MyVpcMyVPCIGW6C595E3B",
        },
        "RouteTableId": {
          "Ref": "MyVpcMyVPCapplicationsubnetSubnet1RouteTable23F4EBCD",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "MyVpcMyVPCapplicationsubnetSubnet1RouteTable23F4EBCD": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/MyVpc/MyVPC/application-subnetSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "MyVpcMyVPC28A4F3FB",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "MyVpcMyVPCapplicationsubnetSubnet1RouteTableAssociationC4106D18": {
      "Properties": {
        "RouteTableId": {
          "Ref": "MyVpcMyVPCapplicationsubnetSubnet1RouteTable23F4EBCD",
        },
        "SubnetId": {
          "Ref": "MyVpcMyVPCapplicationsubnetSubnet1Subnet54B065C1",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "MyVpcMyVPCapplicationsubnetSubnet1Subnet54B065C1": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.0/24",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "application-subnet",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "TestStack/MyVpc/MyVPC/application-subnetSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "MyVpcMyVPC28A4F3FB",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "MyVpcMyVPCapplicationsubnetSubnet2DefaultRoute76B5A664": {
      "DependsOn": [
        "MyVpcMyVPCVPCGW75395802",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "MyVpcMyVPCIGW6C595E3B",
        },
        "RouteTableId": {
          "Ref": "MyVpcMyVPCapplicationsubnetSubnet2RouteTableF09E6FE4",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "MyVpcMyVPCapplicationsubnetSubnet2RouteTableAssociation567406DD": {
      "Properties": {
        "RouteTableId": {
          "Ref": "MyVpcMyVPCapplicationsubnetSubnet2RouteTableF09E6FE4",
        },
        "SubnetId": {
          "Ref": "MyVpcMyVPCapplicationsubnetSubnet2Subnet1F6D9A0F",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "MyVpcMyVPCapplicationsubnetSubnet2RouteTableF09E6FE4": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/MyVpc/MyVPC/application-subnetSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "MyVpcMyVPC28A4F3FB",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "MyVpcMyVPCapplicationsubnetSubnet2Subnet1F6D9A0F": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            1,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.1.0/24",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "application-subnet",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "TestStack/MyVpc/MyVPC/application-subnetSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "MyVpcMyVPC28A4F3FB",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "MyVpcMyVPCdatabasesubnetSubnet1RouteTable2EC17299": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/MyVpc/MyVPC/database-subnetSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "MyVpcMyVPC28A4F3FB",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "MyVpcMyVPCdatabasesubnetSubnet1RouteTableAssociation0798F2A0": {
      "Properties": {
        "RouteTableId": {
          "Ref": "MyVpcMyVPCdatabasesubnetSubnet1RouteTable2EC17299",
        },
        "SubnetId": {
          "Ref": "MyVpcMyVPCdatabasesubnetSubnet1SubnetCB0EE6CF",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "MyVpcMyVPCdatabasesubnetSubnet1SubnetCB0EE6CF": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.2.0/24",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "database-subnet",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
          {
            "Key": "Name",
            "Value": "TestStack/MyVpc/MyVPC/database-subnetSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "MyVpcMyVPC28A4F3FB",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "MyVpcMyVPCdatabasesubnetSubnet2RouteTableAssociationF2AF5313": {
      "Properties": {
        "RouteTableId": {
          "Ref": "MyVpcMyVPCdatabasesubnetSubnet2RouteTableD7FDD806",
        },
        "SubnetId": {
          "Ref": "MyVpcMyVPCdatabasesubnetSubnet2Subnet6FC8DD14",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "MyVpcMyVPCdatabasesubnetSubnet2RouteTableD7FDD806": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/MyVpc/MyVPC/database-subnetSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "MyVpcMyVPC28A4F3FB",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "MyVpcMyVPCdatabasesubnetSubnet2Subnet6FC8DD14": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            1,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.3.0/24",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "database-subnet",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
          {
            "Key": "Name",
            "Value": "TestStack/MyVpc/MyVPC/database-subnetSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "MyVpcMyVPC28A4F3FB",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
